### DISCUSSION TITLE
Help tuning parameters for positional sensing

### DISCUSSION THREAD

**Discussion by davy39** [corrados/edrumulus#139] [Q&A]
Hello,
I'm trying to adjust parameters to make my mps850 10 inch pad working well with positional sensing #138 

I reach a config that's quite OK for low velocity, but not for high velocity : middle hits are a bit shifted to the edge side, meaning that i cannot  get a very middle hit when hitting hard.

Can someone tell me on which parameters to play to improve that ?

Thanks for your help !

**Comment by corrados** [comment]
That is a good question. Edrumulus basically analyzes the first large detected peak. Maybe for your pad the detection of the very first peak is not working correctly and it locks on the second peak which does not give the correct position information.
I would record the pad with 8 kHz sampling rate and do a positional sensing test with low and high velocity. Then you could analyze the detected peaks in the Octave scripts. That may help to find the cause of the issue.

**Comment by davy39** [comment]
Hi @corrados, thanks for your answer.
I tried to tune mostly pos threshold and sens, and got a good response for low velocity but not for high...
I think that you're right about second peak locking, if I understand well what I see with octave.

![](https://i.imgur.com/PI6ASGN.png)

Maybe someone more experimented than me could have a look at [my signals](https://github.com/davy39/edrumulus/raw/mps850/algorithm/signals/mps850_meshpad_10.wav).

What can I change to make it lock on first peak ?

**Reply by corrados** [reply]
Thanks for sharing your signal. That looks really strange. You don't see a distinct peak at the beginning. It looks a bit like noise. Just for comparison, this is what my Roland mesh pad looks like:
![grafik](https://github.com/corrados/edrumulus/assets/46655886/97be4cb7-92db-4c0d-aaf9-39b1afc0fc97)
You can see a clean first peak (blue trace) and the green marker has correctly detected it.
Such a clean first peak is required to get a good positional sensing result. I am really wondering what causes this noise. Is it the type of mesh head which causes some strange resonances? I fear for such a bad signal, my signal processing for positional sensing will not work :-(

**Reply by 3hhh** [reply]
I currently don't have time to look at the signals. However if it was a strong hit, you may just be seeing a signal that was cut off by diodes in your sound card.

You need to understand that a piezo output signal is at +-10V, a sound card input at +-3V3 (IIRC) and the edrumulus circuit input is at +-1V65 (everything else is cut off). Since the input signal range is not compressed (the edrumulus circuit is just a simple voltage divider), there'll always have to be a decision between good response at low velocity or good response at high velocity hits.
If you want the latter, the easiest way may be to put softer foam between the piezo and the mesh head. This dampens the piezo output signal so that you'll get a weaker output signal for strong hits and that signal is hopefully no longer cut off by the diodes. 

Personally I decided to put the [MPS750X piezo back to its original place at the top](https://github.com/corrados/edrumulus/discussions/80#discussioncomment-7188287) as it has a similar effect (piezo further away = less strong signal) and my personal circuit is so sensitive that directly hitting the piezo essentially causes a mute. Obviously pos sensing doesn't work too well then though.

If you happen to have an oscilloscope, you should also be able to measure the real output signal of your piezo.

**Comment by davy39** [comment]
Hi,
Thanks for answer. That's actually a raw recording of piezo, without any analog circuit.
Probably, the problem was that I switched left and right channels while recording : is that better like this ?
That's a plot with default pd120 parameters.

![](https://i.imgur.com/ABEiiGB.png)

**Reply by corrados** [reply]
That looks much better. So, in your first signal, you have captured the rim piezo and not the center-mounted piezo on the mesh head. Could you share a wav of this signal, too?

**Comment by davy39** [comment]
I've captured both of them in the first place, but not in the right order.
Changed it by :
`ffmpeg -i mps850_meshpad_10.wav -af "pan=stereo|c0=c1|c1=c0" output.wav`

I've just replaced the wrong file in [my fork](https://github.com/davy39/edrumulus/commit/29a3d35fca299b05d5baf5e00fce2da45c7832da).

**Reply by corrados** [reply]
Thanks. The positional sensing performance does not look too bad to me:
![grafik](https://github.com/corrados/edrumulus/assets/46655886/7b63009e-131d-4be5-9fac-4a90447cc60b)
I am looking at the right burst (loud hits). The effect that the black markers in the middle are higher than expected I have also seen with my pads. This seems to be an issue of hitting the piezo directly (hot spot effect). I have not found a solution for that problem yet. One method to make positional sensing performance similar for low level and high level strikes is to adjust the threshold/sense parameters. I usually set these parameters that I do not use the entire dynamic but clip at the low/high positions. That works fine for me.

**Reply by corrados** [reply]
Here is a signal from the PD80R pad. The issue in the center is marked with red circles:
![grafik](https://github.com/corrados/edrumulus/assets/46655886/e9876778-fe25-474e-b0bd-7dde5d48c503)
For my pad this is just a single strike. Maybe your piezo is larger than the one on the PD80R.

**Comment by davy39** [comment]
Sorry I didn't answer sooner to thank you, I guess we are facing the same issue then.
I tried modifying resistors values, but still got the same results.
Have you got some advises on how to work on that programmatically ?

**Reply by corrados** [reply]
Changing resistors will not solve that problem. There might be possibilities to solve this in software. I already did some tests but failed to correctly detect hot spot strikes. If the detection of hot spots would work fine, we could adjust the positional sensing value to the middle of the pad since at that place the piezo is located.

**Reply by corrados** [reply]
I guess I have the answer why I do not have issues with positional sensing if I hit hard in the center of my pad. I am using a Roland PD85 mesh pad which has a high output level. Caused by the high level, the ESP32 input is overloaded on hard strikes. For this, I have implemented a solution in software to compensate the clipping. In case I detect a clipping, I compensate it and also I set the position to the middle of the pad, assuming that clipping only happens if you strike close to the piezo. Therefore, the clipping compensation may mask the issue you see with your positional sensing...

**Reply by 3hhh** [reply]
> Caused by the high level, the ESP32 input is overloaded on hard strikes. For this, I have implemented a solution in software to compensate the clipping.

For Piezos clipping more on the negative end, the clipping will not be detected anyway as the diodes pull the voltage down to zero.
At least that's how I explain to me why that clipping solution is currently not working for me: Whenever I hit the piezo, edrumulus goes entirely mute. That's one of the reasons why I removed it from the center...

**Reply by 3hhh** [reply]
In any case it's probably a good idea to optionally disable that clipping compensation.

**Reply by corrados** [reply]
> Whenever I hit the piezo, edrumulus goes entirely mute. 

This is not caused by the clipping compensation. There is another effect causing this. I had such an effect with my kick pedal. When I hit hard, first, the foot punched on the peal and then after a very short period of time the mallet hit the piezo. The foot punch was detected by Edrumulus as a strike and caused by the mask time, the actual mallet hit was masked. This felt like the kick was not detected by Edrumulus at all. I solved this issue in software: https://github.com/corrados/edrumulus/blob/498634ce4e4da4f97486fd742cf2bd6d79b343ff/edrumulus.cpp#L789

**Reply by 3hhh** [reply]
> This is not caused by the clipping compensation.

Hm, I'm not so sure about that as I was talking about snare & toms and doubt that the piezo is so sensitive to detect my hand motion while they are still flying
(what you were describing is essentially the knock on the floor being faster than the mallet, which only makes sense for pedals imho).

I guess I'll perform some tests to identify the root cause when I have some spare time.

**Reply by corrados** [reply]
To debug this issue, you should record data and analyze this effect in Octave. Then you will see what is actually causing the issue.

**Reply by 3hhh** [reply]
![direct_piezo](https://github.com/corrados/edrumulus/assets/13033030/2db868c3-74e3-4ef3-a1c6-6efa21d85830)

Here's a direct piezo hit at 8kHz sample rate (without the middle data):
As expected, the voltage oscillates between 0V and 3.3V for some time before it decays. Since edrumulus has its 0-point at 1.65V, it should be quite straightforward to detect this as a strong hit.
Looks like I'll have to debug this with edrumulus code running.

**Reply by corrados** [reply]
Would be interesting to see what happens some time before this hard hit. You started your plot right where the strike happens. Do you have a signal recorded where a bit of signal is recorded before the strike happens?

**Reply by 3hhh** [reply]
My code recorded all samples except for the 1.4 - 1.5V range (the voltage divider oscillated in that range). So the first sample outside of that range was the ~0.9V you see in the picture.
Btw it did also count the samples incl. the ones in the 1.4 - 1.5V range, i.e. the sample length of the signal is accurate.
It was directly recorded from a Teensy 4.1.

**Reply by corrados** [reply]
Depends on where your threshold is configured. E.g., for the kick trigger example I mentioned:
![grafik](https://github.com/corrados/edrumulus/assets/46655886/d06e0624-ee01-4578-b0ac-7871f3c5af33)
You can see that there is a very small almost unnoticeable peak right before the actual high peak which was detected by Edrumulus. For this special case I needed to implement a fix in software.

Please also note that Edrumulus uses some filtering to detect peaks. Therefore, it may detect peaks in the unfiltered signal which could look like noise. Maybe the peak which was detected by Edrumulus lies in that 1.4-1.5V range which you exclude.

**Reply by 3hhh** [reply]
I see. However then edrumulus would output a low volume MIDI note I guess, but in my case it doesn't output anything at all.

I also noticed that there's a 50% chance that edrumulus will correctly identify a strong hit on direct piezo contact, i.e. I'm not sure whether my above signal is representative.

I guess I'll have to add some debug code to edrumulus to more thoroughly investigate this...

**Reply by 3hhh** [reply]
Btw I also recorded 5 very hard hits (1-2 had undesired rebound) on the piezo. It [looks fine to me](https://github.com/corrados/edrumulus/files/15081660/direct_piezo.zip). So I guess I'll have to stick to my previous debug idea...

(I had to zip the file as github doesn't allow wav files to be uploaded.)

**Reply by corrados** [reply]
Thanks for the recordings. To sum up: Hard strikes on the piezo do not give you any MIDI signal from Edrumulus. Since this is the case, neither my guess for a low level peak right before the main peak nor your guess that clipping compensation is causing this is true since for both you will get at least one MIDI signal back. In my case it would be the MIDI note of the small peak and in case of clipping compensation you get a MIDI note back and the clipping compensation just modifies the velocity value.
Would be very interesting to see what is actually causing these MIDI mutes. Maybe not so easy to find the real cause...

**Reply by 3hhh** [reply]
I opened https://github.com/corrados/edrumulus/issues/142 for further discussion.

**Comment by davy39** [comment]
Thanks for your answer. 
Until now, I've always chosen a resistor value to avoid clipping. So, next step will be tying to decrease resistor value in order to reach the clipping detection when hitting hard.

Could you explain to me how a hot spot affect a recording ? Is it about non-linearity of the signal level ?

**Reply by corrados** [reply]
The hot spot effect happens with professional drum modules, too, see, e.g., [this discussion](https://www.vdrums.com/forum/general/products/1141055-pd-128-snare-any-way-to-fix-hot-spot-in-the-centre-please). It is not about non-linearity but you have a much higher peak if you hit the piezo directly. E.g. if I use the PD-85 pad with my Roland TD-27 module and I strike with same velocity around the middle of the pad, sometimes I get sounds back from the module which are much too loud which feels unnatural.

Roland fixes that problem by using multiple piezos like in the PD-140DS. Others like drum-tec tries to fix this by also using multiple piezos but these pads can be used with a conventional analog input: https://drum-tec-series.de/de/pro (see section "_REVOLUTIONARY NON-HOTSPOT TRIGGERSYSTEM"_).
